File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    answer = []
    arr = [[0 for _ in range(n)] for _ in range(n)]
    start = (0, 0)
    exit = 0
    for i in range(1, n+1):
        exit += i

    def dfs(start):

        stack = [start]

        dir = [(1, 0), (0, 1), (-1, -1)]
        dir_idx = 0
        now_dir = dir[dir_idx]
        change_dir = {0: 1, 1: 2, 2: 0}
        arr[start[0]][start[1]] = 1

        while stack:
            x, y = stack.pop()

            if arr[x][y] == exit:
                return

            dn, dm = now_dir[0], now_dir[1]
            dnx, dmy = dn + x, dm + y
            if 0 <= dnx < n and 0 <= dmy < n:
                if arr[dnx][dmy] == 0:
                    arr[dnx][dmy] = arr[x][y] + 1
                    stack.append((dnx, dmy))
                    continue

            dir_idx = change_dir[dir_idx]
            now_dir = dir[dir_idx]
            stack.append((x, y))

    dfs(start)

    for i in arr:
        for k in i:
            if k != 0:
                answer.append(k)

    return answer

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
    var answer = [];
    let arr = Array.from({length : n}, () => new Array(n).fill(0))
    let  exit = 0
    for (let i = 1; i <= n; i++) {
        exit += i
    }
    
    function dfs(sn, sm) {
        
        arr[sn][sm] = 1
        let stack = [[sn, sm]]
        let dir_idx = 0
        const dir = [[1, 0], [0, 1], [-1, -1]]
        const dir_chan = {0 : 1, 1: 2, 2: 0}
        
        while (stack.length) {
            const [x, y] = stack.pop()
            const [dn, dm] = dir[dir_idx]
            
            if (arr[x][y] === exit) {
                return 
            }
            
            const xn = x + dn
            const ym = y + dm
            
            if ( 0 <= xn && xn < n && 0<= ym && ym < n) {
                if (arr[xn][ym] === 0) {
                arr[xn][ym] = arr[x][y] + 1
                stack.push([xn, ym])
                continue
                }
            }
            dir_idx = dir_chan[dir_idx]
            stack.push([x, y])
        }
    }
    dfs(0, 0)
    
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (arr[i][j] > 0) {
                answer.push(arr[i][j])
            }
        }
    }
    return answer;
}